### PR TITLE
Make avpipe_init() nonblocking

### PR DESCRIPTION
**Pull request by elv-reza**
- This will keep the API and modifies the implementation under the hood.
- Move blocking code from avpipe_init() into avpipe_xc(). This will allow to obtain the handle for cancellation.
- Add some unit tests for the hanging cases with SRT and RTMP.

Nate will continue to enhance corner cases and memory cleanup.

### CODE REVIEW

File: avpipe.c [hunk 1]
```diff
@@ -387,7 +390,7 @@ udp_in_closer(
 
     int fd = *((int64_t *)inctx->opaque);
     int sockfd = *((int *)((int64_t *)inctx->opaque+1));
-    elv_dbg("IN CLOSE UDP fd=%d, sockfd=%d, url=%s\n", fd, sockfd, inctx->url);
+    elv_dbg("IN CLOSE UDP fd=%d, sockfd=%d, url=%s\n", fd, sockfd, inctx->url ? inctx->url : "bogus.mp4");
```
Comment (elv-nate) [inline avpipe.c:393]: This seems to be the issue for inctx not being null, is that it will not necessarily have a URL.

**Comment by elv-nate** [comment]
@elv-reza With regards to the synchronization from before, I ended up deciding that:
- Calling `xc_run` or `avpipe_xc` multiple times with the same context concurrently can just be UB, it's pretty easy to avoid without synchronization.
- `volatile` is sufficient for the `cancelled` value in the en/decoding contexts, as it starts at 0 and after that is only ever set to 1, so multiple writers isn't a real concern.

**Comment by elv-nate** [comment]
All tests ran for me, except for the SRT live tests which timed out for some reason. I will try re-running the SRT live test and see if I can get it to work.

**Review by elv-nate — APPROVED** [review]
